src/components/TodoItem.tsx

```diff
@@ -0,0 +1,185 @@
+  todo: Todo;
+  isTemp?: boolean;
+  onDelete: (todoId: number) => void;
+  isDeleting?: boolean;
+  onToggleStatus: (todoId: number, newStatus: boolean) => void;
+  isLoading?: boolean;
+  loading?: boolean;
+  editingTodoId: number | null;
+  setEditingTodoId: (id: number | null) => void;
+  onRename: (todoId: number, newTitle: string) => void;
+  setError: (message: string) => void;
+  setLoading: React.Dispatch<React.SetStateAction<boolean>>;
+};
+
+export const TodoItem: React.FC<Props> = ({
+  todo: { completed, id, title },
+  isTemp = false,
+  onDelete,
+  isDeleting,
+  onToggleStatus,
+  isLoading,
+  loading,
+  editingTodoId,
+  setEditingTodoId,
+  onRename,
+  setError,
+  setLoading,
+}) => {
+  const isEditing = editingTodoId === id;
+  const [editedTitle, setEditedTitle] = useState(title);
+  const editInputRef = useRef<HTMLInputElement | null>(null);
+
+  useEffect(() => {
+    // коли входимо в режим редагування — фокус на локальний інпут
+    if (isEditing) {
+      // невелике таймаут, щоб елемент встиг промальоватись у DOM
+      setTimeout(() => editInputRef.current?.focus(), 0);
+    }
+  }, [isEditing]);
+
+  useEffect(() => {
+    setEditedTitle(title);
+  }, [title]);
+
+  const handleTitleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    setEditedTitle(event.target.value);
+  };
+
+  const handleEditSubmit = async (): Promise<boolean> => {
+    const trimmedTitle = editedTitle.trim();
+
+    if (!trimmedTitle) {
+      try {
+        setLoading(true);
+        await onDelete(id);
+
+        return true;
+      } catch {
+        setError('Unable to delete a todo');
+        setTimeout(() => setError(''), 3000);
+
+        return false;
+      } finally {
+        setLoading(false);
+      }
+    }
+
+    if (trimmedTitle === title) {
+      setEditingTodoId(null);
+
+      return true;
+    }
+
+    try {
+      setLoading(true);
+      await onRename(id, trimmedTitle);
+
+      return true;
+    } catch {
+      setError('Unable to update a todo');
+
+      return false;
+    } finally {
+      setLoading(false);
+    }
+  };
+
+  const handleKeyUp = async (event: React.KeyboardEvent<HTMLInputElement>) => {
+    if (event.key === 'Escape') {
+      setEditedTitle(title);
+      setEditingTodoId(null);
+    }
+  };
+
+  return (
+    <div
+      data-cy="Todo"
+      className={`todo ${completed ? 'completed' : ''}`}
+      key={id}
+    >
+      <label className="todo__status-label" htmlFor={`todo-status-${id}`}>
+        {/* {} */}
+        <input
+          id={`todo-status-${id}`}
+          data-cy="TodoStatus"
+          type="checkbox"
+          className="todo__status"
+          checked={completed}
+          onChange={() => {
+            onToggleStatus(id, !completed);
+          }}
+        />
+      </label>
+
+      {isEditing ? (
+        <form
+          onSubmit={async e => {
+            e.preventDefault();
+            const ok = await handleEditSubmit();
+
+            if (ok) {
```

Comment:
handleEditSubmit should accept event as argument, do ` e.preventDefault();` inside of the function

Instead of returning `true` in `try` block just set correct state value there